gossip/infostore.go

```diff
@@ -146,26 +152,29 @@ func (is *infoStore) getInfo(key string) *info {
 // addInfo adds or updates an info in the infos map.
 //
 // Returns nil if info was added; error otherwise.
-func (is *infoStore) addInfo(key string, i *info) error {
+func (is *infoStore) addInfo(key string, i *Info) error {
 	// Only replace an existing info if new timestamp is greater, or if
 	// timestamps are equal, but new hops is smaller.
 	var contentsChanged bool
 	if existingInfo, ok := is.Infos[key]; ok {
-		if i.Timestamp < existingInfo.Timestamp ||
-			(i.Timestamp == existingInfo.Timestamp && i.Hops >= existingInfo.Hops) {
+		if iNanos, existingNanos := i.Value.Timestamp.WallTime, existingInfo.Value.Timestamp.WallTime; iNanos < existingNanos ||
+			(iNanos == existingNanos && i.Hops >= existingInfo.Hops) {
 			return util.Errorf("info %+v older than current info %+v", i, existingInfo)
 		}
-		contentsChanged = !reflect.DeepEqual(existingInfo.Val, i.Val)
+		contentsChanged = !gogoproto.Equal(&i.Value, &existingInfo.Value)
```

Comment:
This is wrong, it should only compare the value itself, not the timestamps. Or perhaps now the everything in proto.Value except for the timestamp.